keps/sig-scheduling/1923-try-nominated-node-first/README.md

```diff
@@ -0,0 +1,157 @@
+running on each of the nodes in the cluster to release enough room for the
+pod to be scheduled, the selected node for the incoming pod after the
+preemption will be set as the value of `pod.Status.NominatedNodeName`.
+The nominated node here implies the node that is selected for the incoming
+pod to be scheduled.
+
+Pod scheduling is split into two phases, the scheduling cycle and the binding
+cycle, the scheduling cycle primarily includes predicate and scroing.
+
+When the preemption is happened, the phase of predicate for the nominated node
+has been properly measured, there is a high chance the selected nominated node
+is still the *only* one candidate for the incoming pod to be scheduled.
+
+This KEP proposes to validate the availability of the nominated node directly
+in the new scheduling cycle before looping all over the nodes in the cluster
+for the phase of the predicate.
+
+## Motivation
+
+In real production environment, pods can have different priorites due to business
+needs, the preemption could happen to make sure high priority pods could got scheduled
+and spawned.
+
+For the cluster with huge number of computing nodes, looping over all the nodes in
+the cluster for the phase of the predicate is an unnegligible consuming for scheduler,
+besides, it's unpredicatable that how the out-of-tree plugins are defined, whether
+they are time consuming, this is also a factor that may increase the time for the
+phase of predicate.
+
+### Goals
+
+- In the case of `pod.Status.NominatedNodeName` is set for the incoming pod, scheduler
+  will try the nominated node first for the phase of predicate.
+- Failed to schedule on the nominated node will fall back to verify all the nodes in
+  the cluster instead of preemption again.
+- A new feature gate will be provided for the operator to opt in/out this feature.
+
+## Proposal
+
+### User Stories (Optional)
+
+As a premium user, I want my high priority application got spin up more quickly,
+I don't care where the application is running, whether there is any other nodes
+in the cluster have more disk space, memory size etc., as long as my application
+run normally.
+
+### Notes/Constraints/Caveats (Optional)
+
+Enables this feature also implies the preemptor pod might not be dispatched
+to the best candidated node in some corner case, e.g. another node releases
+the resources and becomes the best candidate while the victim pods got removed
+from the nominated node, but enable this feature will make scheduler to consider
+the nominated node firstly.
+
+## Design Details
+
+### Implementation Details
+
+1. In the phase of predicate, which is currently implemented in the method of `findNodesThatFitPod`,
+   check the nominated directly if the incoming pod has the `pod.Status.NominatedNodeName` defined
+   and the feature gate is enabled.
+2. In case the nominated node is not suit the incoming pod anymore, return `fmt.Errorf` instead of
```

Comment:
This shouldn't be like that.

Let's say the nominated Node had 2 Pods preempted and one was removed. At this point, kube-scheduler might try to schedule the Pod and fail. But in this case we should keep trying.

Also, errors are supposed to signal internal errors, meaning that something went wrong internally. This doesn't qualify.

Reply:
Additionally, if the Pod doesn't fit in the nominated Node, aren't we supposed to iterate over the rest of the Nodes? Please add those details.

Reply:
Maybe we can go onto a preemption evaluation cycle again and try the nominated Node first. There is a chance that the same Node will still be the nominated one. But if not, it gets changed.

Reply:
Hi Aldo,  I seriously considered those approaches and add a section to explain why it is designed as this, pls take a look and like to see your feedback.